runtime/vm/UpcallThunkMem.cpp

```diff
@@ -87,66 +87,100 @@ subAllocateThunkFromHeap(J9UpcallMetaData *data)
 {
 	J9JavaVM * vm = data->vm;
 	UDATA thunkSize = data->thunkSize;
+	J9UpcallThunkHeapList *thunkHeapHead = vm->thunkHeapHead;
+	J9UpcallThunkHeapList *thunkHeapNode = NULL;
+	J9UpcallThunkHeapWrapper *thunkHeapWrapper = NULL;
 	J9Heap *thunkHeap = NULL;
 	void *subAllocThunkPtr = NULL;
+	bool succeeded = false;
 	PORT_ACCESS_FROM_JAVAVM(vm);
 
 	Trc_VM_subAllocateThunkFromHeap_Entry(thunkSize);
 
-	if (NULL == vm->thunkHeapWrapper) {
-		thunkHeap = allocateThunkHeap(data);
-		if (NULL == thunkHeap) {
+	if (NULL == thunkHeapHead) {
+retry:
+		thunkHeapHead = allocateThunkHeap(data);
+		if (NULL == thunkHeapHead) {
 			/* The thunk address is NULL if VM fails to allocate the thunk heap */
 			goto done;
 		}
-	} else {
-		thunkHeap = vm->thunkHeapWrapper->heap;
 	}
+	thunkHeapNode = thunkHeapHead;
+
+	while (!succeeded) {
+		thunkHeapWrapper = thunkHeapNode->thunkHeapWrapper;
+		thunkHeap = thunkHeapWrapper->heap;
 
 #if defined(OSX) && defined(AARCH64)
-	pthread_jit_write_protect_np(0);
+		pthread_jit_write_protect_np(0);
```

Comment:
Although it's not a feature introduced by this change, I don't think calls to `pthread_jit_write_protect_np()` should be in the loop. Also, there are paths that bypass the call `pthread_jit_write_protect_np(1)`.

Reply:
@knn-k, please double-check whether `pthread_jit_write_protect_np()` should be in the loop or not. Will move out if unnecessary.

Reply:
I just moved them out of the loop and added one more at `goto retry` to ensure it works in the correct way given `pthread_jit_write_protect_np(0)` and pthread_jit_write_protect_np(1) are intended for `j9heap_allocate` & `j9heap_free` on `macOS/Aarch64`.

@knn-k, please help check whether all paths are covered by these adjustments.